lntest/itest/lnd_payment_test.go

```diff
@@ -511,196 +511,94 @@ func testBidirectionalAsyncPayments(net *lntest.NetworkHarness, t *harnessTest)
+	// should be dispatched shortly below.
+	req := &lnrpc.InvoiceSubscription{}
+	bobInvoiceSubscription := bob.RPC.SubscribeInvoices(req)
+
 	// Open a channel with 500k satoshis between Alice and Bob with Alice
 	// being the sole funder of the channel.
-	chanPoint := openChannelAndAssert(
-		t, net, net.Alice, net.Bob,
-		lntest.OpenChannelParams{
-			Amt: chanAmt,
-		},
+	chanPoint := ht.OpenChannel(
+		alice, bob, lntemp.OpenChannelParams{Amt: chanAmt},
 	)
 
 	// Next create a new invoice for Bob requesting 1k satoshis.
-	// TODO(roasbeef): make global list of invoices for each node to re-use
-	// and avoid collisions
 	const paymentAmt = 1000
 	invoice := &lnrpc.Invoice{
 		Memo:      "testing",
-		RPreimage: makeFakePayHash(t),
+		RPreimage: ht.Random32Bytes(),
 		Value:     paymentAmt,
 	}
-	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
-	invoiceResp, err := net.Bob.AddInvoice(ctxt, invoice)
-	if err != nil {
-		t.Fatalf("unable to add invoice: %v", err)
-	}
+	invoiceResp := bob.RPC.AddInvoice(invoice)
 	lastAddIndex := invoiceResp.AddIndex
 
-	// Create a new invoice subscription client for Bob, the notification
-	// should be dispatched shortly below.
-	req := &lnrpc.InvoiceSubscription{}
-	ctx, cancelInvoiceSubscription := context.WithCancel(ctxb)
-	bobInvoiceSubscription, err := net.Bob.SubscribeInvoices(ctx, req)
-	if err != nil {
-		t.Fatalf("unable to subscribe to bob's invoice updates: %v", err)
-	}
-
-	var settleIndex uint64
-	quit := make(chan struct{})
-	updateSent := make(chan struct{})
-	go func() {
-		invoiceUpdate, err := bobInvoiceSubscription.Recv()
-		select {
-		case <-quit:
-			// Received cancellation
-			return
-		default:
-		}
-
-		if err != nil {
-			t.Fatalf("unable to recv invoice update: %v", err)
-		}
-
-		// The invoice update should exactly match the invoice created
-		// above, but should now be settled and have SettleDate
-		if !invoiceUpdate.Settled { // nolint:staticcheck
-			t.Fatalf("invoice not settled but should be")
-		}
-		if invoiceUpdate.SettleDate == 0 {
-			t.Fatalf("invoice should have non zero settle date, but doesn't")
-		}
-
-		if !bytes.Equal(invoiceUpdate.RPreimage, invoice.RPreimage) {
-			t.Fatalf("payment preimages don't match: expected %v, got %v",
-				invoice.RPreimage, invoiceUpdate.RPreimage)
-		}
-
-		if invoiceUpdate.SettleIndex == 0 {
-			t.Fatalf("invoice should have settle index")
-		}
-
-		settleIndex = invoiceUpdate.SettleIndex
-
-		close(updateSent)
-	}()
-
-	// Wait for the channel to be recognized by both Alice and Bob before
-	// continuing the rest of the test.
-	err = net.Alice.WaitForNetworkChannelOpen(chanPoint)
-	if err != nil {
-		// TODO(roasbeef): will need to make num blocks to advertise a
-		// node param
-		close(quit)
-		t.Fatalf("channel not seen by alice before timeout: %v", err)
-	}
+	// With the above invoice added, we should receive an update event.
+	invoiceUpdate := ht.ReceiveInvoiceUpdate(bobInvoiceSubscription)
```

Comment:
You mean that if we subscribe too late we'd miss an event? Some of the subscriptions would give you missed events, but not all of them. Initially I event thought about creating a list of subscribers when the node is started so we never miss events, like how we handle topology updates. Later on I thought it would be better to not place too much magic in the test but let the subscription change its behavior to simulate real-life conditions.